src/ephy-sync-crypto.c

```diff
@@ -0,0 +1,952 @@
+/* -*- Mode: C; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/*
+ *  Copyright © 2016 Gabriel Ivascu <ivascu.gabriel59@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+#include "ephy-sync-crypto.h"
+
+#include "ephy-sync-utils.h"
+
+#include <glib/gstdio.h>
+#include <libsoup/soup.h>
+#include <nettle/aes.h>
+#include <string.h>
+
+#define HAWK_VERSION  1
+
+static const char hex_digits[] = "0123456789abcdef";
+
+EphySyncCryptoHawkOptions *
+ephy_sync_crypto_hawk_options_new (const char *app,
+                                   const char *dlg,
+                                   const char *ext,
+                                   const char *content_type,
+                                   const char *hash,
+                                   const char *local_time_offset,
+                                   const char *nonce,
+                                   const char *payload,
+                                   const char *timestamp)
+{
+  EphySyncCryptoHawkOptions *options;
+
+  options = g_slice_new (EphySyncCryptoHawkOptions);
+  options->app = g_strdup (app);
+  options->dlg = g_strdup (dlg);
+  options->ext = g_strdup (ext);
+  options->content_type = g_strdup (content_type);
+  options->hash = g_strdup (hash);
+  options->local_time_offset = g_strdup (local_time_offset);
+  options->nonce = g_strdup (nonce);
+  options->payload = g_strdup (payload);
+  options->timestamp = g_strdup (timestamp);
+
+  return options;
+}
+
+static EphySyncCryptoHawkArtifacts *
+
```

Comment:
Please rearrange these functions so that the new function is directly above the corresponding free function; that would be easier to read.